File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.diameter=0
        def findDiam(root):
            if root is None:
                return 0
            left=findDiam(root.left)
            right=findDiam(root.right)
            self.diameter=max(self.diameter,(left+right))
            return max(left,right) +1
        findDiam(root)
        return self.diameter

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxD = 0;
    public int helper(TreeNode root){
            if (root==null){
                return 0;
            }
            int left = helper(root.left);
            int right = helper(root.right);
            maxD = Math.max(maxD,(left+right));
            return 1 + Math.max(left,right);
        }
    public int diameterOfBinaryTree(TreeNode root) {
        maxD = 0;
        helper(root);
        return maxD;
    }
}